marlowe-playground-client/src/Simulation.purs

```diff
@@ -137,10 +139,16 @@ handleAction _ (SetEditorText contents) = do
   editorSetValue contents
   updateContractInState contents
 
+handleAction _ StartSimulation = do
+  assign (_currentMarloweState <<< _executionState) (Just $ emptyExecutionStateWithSlot zero)
+  moveToSlot zero
+
 handleAction _ (MoveSlot slot) = do
-  currentSlot <- use (_currentMarloweState <<< _slot)
+  maybeExecutionState <- use (_currentMarloweState <<< _executionState)
+  let
+    slotGTcurrentSlot = maybe false (\x -> slot > (x ^. _slot)) maybeExecutionState
```

Comment:
I'm not a big fan of `maybe` and in this case I think you can fit it all nicely in the lens bit, something like
```purescript
is (_currentMarloweState <<< _executionState <<< _Just <<< _slot <<< to (greaterThan slot)) state`
```
`is` and `has` can be very nice.

Reply:
ah, forgot this is getting it from the state monad, I don't think there's an equivalent of `use` for `has` and `is`. Still the following might be nicer
```purescript
mSlotGTcurrentSlot  <- use (_currentMarloweState <<< _executionState <<< _Just <<< _slot <<< to (greaterThan slot)) state
when (fromMaybe false mSlotGTcurrentSlot)
```
but maybe not. Not sure now

Reply:
actually to take this further, it might be really nice to add to https://github.com/input-output-hk/plutus/blob/master/web-common/src/Data/Lens/Extra.purs functions for `is` and `has` that get their data from the monad, just like `peruse`. However they might be hard to name. Would be very useful for things like this though.

Reply:
looking at other parts of this PR I think at a minimum it's best to move `(\x -> slot > (x ^. _slot))` into a `_Just <<< _slot` part and have something like `maybe false (greaterThan slot) mSlot`

Reply:
It doesn't seem to work, I am guessing it is because Slot is not a semiring or a monoid or something like that. For the last one, I don't think having slot be in a Maybe by itself makes much sense, because we are supposed to have a slot if and only if the execution is running

Reply:
So first thing you can do is
```purescript
  maybeInTheFuture <- peruse (_currentMarloweState <<< _executionState <<< _Just <<< _slot <<< to (lessThan slot))
  let
    inTheFuture = fromMaybe false maybeInTheFuture
  significantSlot <- use (_marloweState <<< _Head <<< to nextSignificantSlot)
  when inTheFuture do
```
but we can do better than that if we add use the `has` function.
```purescript
  s <- get
  let
    inTheFuture = has (_currentMarloweState <<< _executionState <<< _Just <<< _slot <<< to (lessThan slot)) s
  significantSlot <- use (_marloweState <<< _Head <<< to nextSignificantSlot)
  when inTheFuture do
```
Here we just want to get rid of the `get` and add the `MonadState` equivalent of `is` to `Data.Lens.Extra`, lets call it `stateHas` for now, although the name could be better.
```purescript
  inTheFuture <- stateHas (_currentMarloweState <<< _executionState <<< _Just <<< _slot <<< to (lessThan slot))
  significantSlot <- use (_marloweState <<< _Head <<< to nextSignificantSlot)
  when inTheFuture do
```
I think if you get your head around `view` and `preview`, `use` and `peruse` and `has` and `is` then you can go a long way improving the readability of the code that uses deeply nested `MonadState`.

Reply:
noice :)
I have no idea what `Determines whether a Fold has at least one focus.` means, but it sounds like `has` will return `true` if we get `Just` and `false` if we get `Nothing`, as opposed to return whatever is inside the `Just`, which is what we would need :/

Reply:
yes I think the documentation is not very clear. I've used it in the past to do this but please check. I still get confused with `has` and `is`

Reply:
no you are right, `has` is wrong

Reply:
ahhhh, sorry what you want is `has (_currentMarloweState <<< _executionState <<< _Just <<< _slot <<< to (lessThan slot) <<< only true) s`

Reply:
or better `is (_currentMarloweState <<< _executionState <<< _Just <<< _slot <<< only (lessThan slot)) s`